### PR TITLE
[GRADIENT] Add backward operator to enable backward graph 

### PR DESCRIPTION
**Pull request by yuruofeifei**
@kevinthesun and I are currently working on supporting backward graph in NNVM.

Our current plan is to add backward to most commonly used operators in top folder.

This PR is trying to gather feedback from community, more operators support will come in the following PR.

### CODE REVIEW

**Comment by kevinthesun** [comment]
With this patch, only elementwise_add and sqrt operator are covered with backward. We think it would be good to have a complete framework first and then add more operators. Same idea apply to potential  parameter update graph pass. Currently there is no optimizer operator defined in nnvm. We can start with naive SGD and add more latter.

File: include/nnvm/node.h [hunk 1]
```diff
@@ -141,7 +141,8 @@ inline NodeEntry MakeNode(
     std::string node_name,
     std::vector<NodeEntry> inputs,
     std::unordered_map<std::string, std::string> attrs =
-    std::unordered_map<std::string, std::string>()) {
+    std::unordered_map<std::string, std::string>(),
+    NodePtr fwd_node = nullptr) {
```
Comment (kevinthesun) [inline include/nnvm/node.h]: Add param documentation?

**Comment by tqchen** [comment]
For most of the new simple operators like sqrt, we prefer not do add them as backward operator (with reference to forward version), but instead define FGradient by returning composition of new forward operators.

Bulk backward operator is only needed when there is need of shape inference like conv2d

**Comment by yuruofeifei** [comment]
@tqchen @ZihengJiang @kevinthesun @piiswrong 

I have addressed Tianqi's comment and added tests for simple operators.

Please let me know your thoughts.

File: tests/python/compiler/test_top_level1.py [hunk 2]
```diff
@@ -56,6 +127,19 @@ def test_log():
         y_np = np.log(data)
         np.testing.assert_allclose(out.asnumpy(), y_np, atol=1e-5, rtol=1e-5)
 
+        # backward
+        graph._set_symbol_list_attr('grad_ys', y)
```
Comment (ZihengJiang) [inline tests/python/compiler/test_top_level1.py]: Consider add a python function for this

Comment (kevinthesun) [reply tests/python/compiler/test_top_level1.py]: I'll work on this to add creating backward graph for both c_api and python api.

**Comment by yuruofeifei** [comment]
Currently, nnvm does not have ones or zeros ops, added fill/fill_like operator to serve the same functionality. 
@kevinthesun and I think current ops are complete for testing backward graph generation flow after this PR. 
Can you please review? @tqchen 
Thanks!

**Comment by tqchen** [comment]
please update docs/top.rst to include the operators. cc @ZihengJiang for doing a code review, i have already done a rough pass

**Comment by yuruofeifei** [comment]
@ZihengJiang @yidawang @piiswrong

**Comment by tqchen** [comment]
there is a fix of tutorial commit at most just now, please rebase

**Comment by yuruofeifei** [comment]
Test passed

**Comment by ZihengJiang** [comment]
LGTM

**Pull request merged**